File: src/detector.rs
//! Feed-agnostic fee-adjusted arb detector — a direct port of the TS version.
//! Feed it price ticks from any source (gRPC now, ShredStream later) and it
//! emits arb open/close with lifetime (slots + ms) = the reaction budget. A
//! "real" arb requires the cross-venue gap to exceed the SUM of both fees.

#[derive(Clone, Debug)]
pub struct Tick {
    pub venue: &'static str,
    pub price: f64, // quote per base (USDC per SOL)
    pub slot: u64,
    pub ts_ms: u128,
}

#[derive(Clone, Debug)]
#[allow(dead_code)] // buy/sell venue + open/close slot consumed by the PR4 report
pub struct ArbEvent {
    pub buy_venue: &'static str,
    pub sell_venue: &'static str,
    pub open_slot: u64,
    pub close_slot: u64,
    pub lifetime_slots: u64,
    pub lifetime_ms: u128,
    pub peak_net_bps: f64,
}

pub enum TickResult {
    None,
    Open { net_bps: f64 },
    Close(ArbEvent),
}

struct OpenState {
    buy: &'static str,
    sell: &'static str,
    open_slot: u64,
    open_ts_ms: u128,
    peak_net_bps: f64,
    last_slot: u64,
    last_ts_ms: u128,
}

pub struct Detector {
    venues: (&'static str, &'static str),
    pub threshold_bps: f64,
    last_a: Option<f64>,
    last_b: Option<f64>,
    open: Option<OpenState>,
}

impl Detector {
    pub fn new(a: &'static str, b: &'static str, fee_a: f64, fee_b: f64) -> Self {
        Self {
            venues: (a, b),
            threshold_bps: fee_a + fee_b,
            last_a: None,
            last_b: None,
            open: None,
        }
    }

    pub fn on_tick(&mut self, t: &Tick) -> TickResult {
        let (a, b) = self.venues;
        if t.venue == a {
            self.last_a = Some(t.price);
        } else if t.venue == b {
            self.last_b = Some(t.price);
        }
        let (pa, pb) = match (self.last_a, self.last_b) {
            (Some(pa), Some(pb)) => (pa, pb),
            _ => return TickResult::None,
        };

        let signed_gap_bps = ((pb - pa) / pa) * 10_000.0; // + => B dearer
        let (net, buy, sell) = if signed_gap_bps > self.threshold_bps {
            (signed_gap_bps - self.threshold_bps, a, b)
        } else if signed_gap_bps < -self.threshold_bps {
            (-signed_gap_bps - self.threshold_bps, b, a)
        } else {
            (0.0, "", "")
        };

        if net > 0.0 {
            match &mut self.open {
                None => {
                    self.open = Some(OpenState {
                        buy,
                        sell,
                        open_slot: t.slot,
                        open_ts_ms: t.ts_ms,
                        peak_net_bps: net,
                        last_slot: t.slot,
                        last_ts_ms: t.ts_ms,
                    });
                    return TickResult::Open { net_bps: net };
                }
                Some(o) => {
                    o.peak_net_bps = o.peak_net_bps.max(net);
                    o.last_slot = t.slot;
                    o.last_ts_ms = t.ts_ms;
                    return TickResult::None;
                }
            }
        }

        if let Some(o) = self.open.take() {
            let close_slot = if t.slot != 0 { t.slot } else { o.last_slot };
            let _ = o.last_ts_ms;
            return TickResult::Close(ArbEvent {
                buy_venue: o.buy,
                sell_venue: o.sell,
                open_slot: o.open_slot,
                close_slot,
                lifetime_slots: close_slot.saturating_sub(o.open_slot),
                lifetime_ms: t.ts_ms.saturating_sub(o.open_ts_ms),
                peak_net_bps: o.peak_net_bps,
            });
        }
        TickResult::None
    }
}

pub fn median_f64(mut v: Vec<f64>) -> f64 {
    if v.is_empty() {
        return 0.0;
    }
    v.sort_by(|x, y| x.partial_cmp(y).unwrap());
    v[v.len() / 2]
}

pub fn median_u128(mut v: Vec<u128>) -> u128 {
    if v.is_empty() {
        return 0;
    }
    v.sort_unstable();
    v[v.len() / 2]
}

File: src/lib/detector.ts
// Port of src/detector.rs
//
// Feed-agnostic fee-adjusted arb detector — a direct port of the TS version.
// Feed it price ticks from any source (gRPC now, ShredStream later) and it
// emits arb open/close with lifetime (slots + ms) = the reaction budget. A
// "real" arb requires the cross-venue gap to exceed the SUM of both fees.

export interface Tick {
  venue: string;
  price: number; // quote per base (USDC per SOL)
  slot: number;
  tsMs: number;
}

export interface ArbEvent {
  buyVenue: string;
  sellVenue: string;
  openSlot: number;
  closeSlot: number;
  lifetimeSlots: number;
  lifetimeMs: number;
  peakNetBps: number;
}

export type TickResult =
  | { kind: 'open'; netBps: number }
  | { kind: 'close'; event: ArbEvent }
  | { kind: 'none' };

interface OpenState {
  buy: string;
  sell: string;
  openSlot: number;
  openTsMs: number;
  peakNetBps: number;
  lastSlot: number;
  lastTsMs: number;
}

export class Detector {
  private readonly venues: [string, string];
  readonly thresholdBps: number;
  private lastA: number | undefined;
  private lastB: number | undefined;
  private open: OpenState | undefined;

  constructor(a: string, b: string, feeA: number, feeB: number) {
    this.venues = [a, b];
    this.thresholdBps = feeA + feeB;
    this.lastA = undefined;
    this.lastB = undefined;
    this.open = undefined;
  }

  onTick(t: Tick): TickResult {
    const [a, b] = this.venues;
    if (t.venue === a) {
      this.lastA = t.price;
    } else if (t.venue === b) {
      this.lastB = t.price;
    }
    const pa = this.lastA;
    const pb = this.lastB;
    if (pa === undefined || pb === undefined) {
      return { kind: 'none' };
    }

    const signedGapBps = ((pb - pa) / pa) * 10_000.0; // + => B dearer
    let net: number;
    let buy: string;
    let sell: string;
    if (signedGapBps > this.thresholdBps) {
      net = signedGapBps - this.thresholdBps;
      buy = a;
      sell = b;
    } else if (signedGapBps < -this.thresholdBps) {
      net = -signedGapBps - this.thresholdBps;
      buy = b;
      sell = a;
    } else {
      net = 0.0;
      buy = '';
      sell = '';
    }

    if (net > 0.0) {
      if (this.open === undefined) {
        this.open = {
          buy,
          sell,
          openSlot: t.slot,
          openTsMs: t.tsMs,
          peakNetBps: net,
          lastSlot: t.slot,
          lastTsMs: t.tsMs,
        };
        return { kind: 'open', netBps: net };
      } else {
        this.open.peakNetBps = Math.max(this.open.peakNetBps, net);
        this.open.lastSlot = t.slot;
        this.open.lastTsMs = t.tsMs;
        return { kind: 'none' };
      }
    }

    if (this.open !== undefined) {
      const o = this.open;
      this.open = undefined;
      const closeSlot = t.slot !== 0 ? t.slot : o.lastSlot;
      return {
        kind: 'close',
        event: {
          buyVenue: o.buy,
          sellVenue: o.sell,
          openSlot: o.openSlot,
          closeSlot,
          lifetimeSlots: Math.max(0, closeSlot - o.openSlot),
          lifetimeMs: Math.max(0, t.tsMs - o.openTsMs),
          peakNetBps: o.peakNetBps,
        },
      };
    }
    return { kind: 'none' };
  }
}

export function medianF64(v: number[]): number {
  if (v.length === 0) return 0.0;
  const sorted = [...v].sort((x, y) => x - y);
  return sorted[Math.floor(sorted.length / 2)];
}

export function medianU128(v: number[]): number {
  if (v.length === 0) return 0;
  const sorted = [...v].sort((x, y) => x - y);
  return sorted[Math.floor(sorted.length / 2)];
}
